### PR TITLE
[TechDocs] Avoid run-time availability checks when instantiating publishers

### PR DESCRIPTION
**Pull request by dhenneke**
When the configured S3 bucket is not available, the publisher can't be created and the Backstage application crashes. This is unfortunate since these kinds of configuration problems can happen and we can't guarantee that the S3 bucket is always available / accessible from the cluster. So not being able to start backstage in production when the S3 config is invalid is not really nice. Especially because the backend can still run nicely even if some selected features might be offline.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by iamEAP — COMMENTED** [review]
Hey @dhenneke!  Thanks for the contribution.

I recognize the annoyance when this kind of thing happens, but I'm not sure this is the right place to alleviate it.  Couple of thoughts off the top of my head:

- `@backstage/techdocs-common` is not only used in TechDocs Backend, but also by the `techdocs-cli` (for which a primary use case is building/publishing docs in CI/CD).  In this situation, a fast and decisive _"This is misconfigured or AWS is unavailable..."_ would be more helpful than letting it continue and trip over itself.
- AWS is but one supported publisher, and I would imagine a similar annoyance would be encounterable in the others.

With that in mind, maybe this is something we let individual adopters decide how to handle by `catch`'ing in their respective app backends?  I think you could achieve similar results by doing something like this in the equivalent file in your Backstage instance:

```diff
diff --git a/packages/backend/src/plugins/techdocs.ts b/packages/backend/src/plugins/techdocs.ts
index 9883f2afb..b24413a71 100644
--- a/packages/backend/src/plugins/techdocs.ts
+++ b/packages/backend/src/plugins/techdocs.ts
@@ -43,10 +43,15 @@ export default async function createPlugin({
   // Publisher is used for
   // 1. Publishing generated files to storage
   // 2. Fetching files from storage and passing them to TechDocs frontend.
-  const publisher = await Publisher.fromConfig(config, {
+  let publisher;
+  try {
+    publisher = await Publisher.fromConfig(config, {
       logger,
       discovery,
     });
+  } catch (e) {
+    return Router();
+  }
 
   // Docker client (conditionally) used by the generators, based on techdocs.generators config.
   const dockerClient = new Docker();
```

**Comment by dhenneke** [comment]
> * `@backstage/techdocs-common` is not only used in TechDocs Backend, but also by the `techdocs-cli` (for which a primary use case is building/publishing docs in CI/CD).  In this situation, a fast and decisive _"This is misconfigured or AWS is unavailable..."_ would be more helpful than letting it continue and trip over itself.

Yes that's a valid point. But the code only checks that the bucket exists, not whether TechDocs is able to write into the Bucket. So the upload and the download code should always include failure checks that give reasonable error messages when something went wrong. I'm not sure if it would be way worse if we don't do this check and still crash as soon as `techdocs-cli publish` tries to upload a file.

Proposal: It could be an option to add e.g. a `validateConnection: Promise<boolean>` function to the `PublisherBase` interface that explicitly runs the check, instead of implicitly doing this in the constructor code of each publisher.
The `techdocs-cli` can call it [here](https://github.com/backstage/techdocs-cli/blob/28a144ca42703aead952b67c2eb842ec06518ba4/packages/techdocs-cli/src/commands/publish/publish.ts#L83) and the `techdocs-backend` could skip it. The check can't distinguish a failed configuration from a temporary connection error which is irrelevant in the CI-based publishing but relevant in the backend. After all, even S3/GCS/… could be offline and this shouldn't take the complete application down. Or it could be that the IAM is messed up for some reason.

> * AWS is but one supported publisher, and I would imagine a similar annoyance would be encounterable in the others.

I hadn't assumed that this is a general pattern but you are right. This could be solved with the proposed decoupling between config initialization and a quick sanity check.

> With that in mind, maybe this is something we let individual adopters decide how to handle by `catch`'ing in their respective app backends? I think you could achieve similar results by doing something like this in the equivalent file in your Backstage instance:

I also thought about this one. But it has no chance to recover from transient failures since techdocs isn't initiated at all. I would prefer to skip the check and fail on the individual operations instead. They could later be guarded by e.g. a circuit breaker to keep being responsive, but this is unrelated to the initial sanity-check since such failures could happen always.

@iamEAP should I give my proposal a try and move the sanity-checks out of the `*.fromConfig(…)` calls?

**Comment by iamEAP** [comment]
Thanks for the additional context @dhenneke, I see what you're getting at.  I agree, the run-time check against the various providers for existence/availability is probably not warranted just to instantiate the class, and it definitely makes sense that any actual errors should be handled on the specific operations.

With that in mind, I was thinking a `validateConfig()` method seemed like overkill to require of a provider (in the abstract), but then I thought it might actually serve a super useful purpose on a healthcheck endpoint (for e.g. monitoring, a kube readiness probe, etc), were one to deploy the TechDocs backend in isolation.

So I like it!  I'd say you're go to move the sanity checks out of the static instantiation methods into another method.

Perhaps it would be good to be more expressive than a boolean?  (e.g. certainly the bucket should exist, but if I'm using the publisher for publishing, I'd want to know that it was writeable, if I'm using the publisher to read docs from the bucket, I'd want to know that it was readable, and I may not require one or the other, depending on context/deployment). Alternately, we could split it into a few checks: `isValid: Promise<boolean>`, `isReadable: Promise<boolean>`, `isWriteable: Promise<boolean>`.

Naturally, we can just start with the first (to solve this issue) and track the others (plus that health endpoint) in a follow-up.

**Comment by dhenneke** [comment]
I went with the following method that can be extended with further checks in the future:
```ts
// ...

/**
 * Result for the validation check.
 */
export type ConfigurationValidationResponse = {
  /** Tells whether the configuration is valid.   */
  isValid: boolean;
};

// ...

export interface PublisherBase {
  /**
   * Check if the configuration is valid. This check tries to perform certain checks to see if the
   * publisher is configured correctly and can be used to publish or read documentations.
   * The different implementations might e.g. use the provided service credentials to access the
   * target or check if a folder/bucket is available.
   */
  validateConfiguration(): Promise<ConfigurationValidationResponse>;

  // ...
}
```

I implemented it in all publishers and added tests for each. Note that I could only test this with AWS in a real environment. I also removed some duplicated tests in the `azureBlobStorage.test.ts`. This can be best viewed when ignoring whitespace changes.

This code is not run at startup anymore. I'm not sure if we want to also advise the users to call it in their `packages/backstage/src/plugins/techdocs.ts` or even call it in the `Publisher.fromConfig(…)` to at least trigger the logging of errors. It could be helpful for users to see them even if they might not want their service to crash. What do you think?

A follow-up would be to call this code from the `techdocs-cli` as soon as this PR is merged and the library is released next Thursday.

File: .changeset/fluffy-lies-complain.md
Comment (iamEAP) [inline .changeset/fluffy-lies-complain.md:2]: With the API interface addition to `PublisherBase`, probably more appropriately a `minor` change, now.

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (iamEAP) [inline packages/techdocs-common/src/stages/publish/awsS3.ts:144]: Now that we may not necessarily _only_ be running this at start-up, it may make sense to tweak these success messages a little bit to be more of a representation of continuous state?  e.g. instead of `Successfully connected to...`, something like `Able to connect to...`

File: packages/techdocs-common/src/stages/publish/types.ts
Comment (iamEAP) [inline packages/techdocs-common/src/stages/publish/types.ts:70]: Now that I'm seeing this in context, and given we may implement additional checks at some later stage, `validateConfiguration()` feels a little misleading.  At a glance, I would imagine "validate configuration" to be more of a quick "does the config match the schema" or "do I have the minimum viable config keys set" type of check rather than a "I'm going to make network calls against an external service" type of check.

Perhaps `PublisherBase::getReadiness(): Promise<ReadinessResponse>`?  Or something similar.

And then in the context of "readiness" I guess instead of `isValid` (which is kind of config-y) could be `isAvailable`?

**Review by iamEAP — COMMENTED** [review]
Thanks @dhenneke!  Looking good.  A couple of suggestions inline.

Regarding:

> _This code is not run at startup anymore. I'm not sure if we want to also advise the users to call it in their `packages/backstage/src/plugins/techdocs.ts` or even call it in the `Publisher.fromConfig(…)` to at least trigger the logging of errors._

Agree, the errors are super helpful when folks are getting started with Backstage.  Given we may extend in the future with other checks, I don't know that it necessarily makes sense to run in `fromConfig()`, but it'd be great if you added a call to it in the example backend in this repo, as well as in the `create-app` equivalent [here](https://github.com/backstage/backstage/blob/master/packages/create-app/templates/default-app/packages/backend/src/plugins/techdocs.ts).

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (dhenneke) [reply packages/techdocs-common/src/stages/publish/awsS3.ts:144]: Tbh, I don't really like emitting logs for the happy-case at all. Sure, they are nice to have for a one-off call, but if this might be used to be run repeatedly as a healthcheck, it is not the task of the logging tool to monitor health checks, but there will be a library/external tool that calls it and should do the logging if needed. I would prefer to leave it as-is for now.

File: packages/techdocs-common/src/stages/publish/types.ts
Comment (dhenneke) [reply packages/techdocs-common/src/stages/publish/types.ts]: 👍  I updated it

**Comment by dhenneke** [comment]
Looks like the E2E tests are broken in all PRs 😕

**Comment by benjdlambert** [comment]
@dhenneke working on it. Think this fixes it #5291

**Review by iamEAP — APPROVED** [review]
Nice!  Thanks for coming along for the journey @dhenneke.  I'm good with this change (perhaps a quick rebase on master will take care of the E2E issue).

Once we get this merged, I'll...
- Create a follow-up issue in `techdocs-cli` to make use of the new readiness check.
- Create a follow-up issue here to add readable/writeable checks and statuses for the existing providers + a health check route that leverages them.

**Comment by dhenneke** [comment]
I added a changeset to bump `create-app`.

@backstage/maintainers can I get a final review for the changes in the default backend and in the `create-app` template?

**Comment by iamEAP** [comment]
One more ping to @backstage/maintainers (the above doesn't appear to have actually made a proper link, somehow) ^^^

**Review by benjdlambert — APPROVED** [review]
+1 for the `create-app` changes. I don't know enough about the techdocs side but @iamEAP seems to have that covered!

**Pull request merged**
